src/main/scala/org/ergoplatform/nodeView/state/UtxoStateReader.scala

```diff
@@ -30,36 +33,48 @@ trait UtxoStateReader extends ErgoStateReader with TransactionValidation[ErgoTra
     * This validation does not guarantee that transaction will be valid in future
     * as soon as state (both UTXO set and state context) will change.
     */
-  override def validate(tx: ErgoTransaction): Try[Unit] =
+  override def validate(tx: ErgoTransaction): Try[Unit] = {
     tx.statelessValidity
-      .flatMap(_ =>
+      .flatMap { _ =>
         tx.statefulValidity(tx.inputs.flatMap(i => boxById(i.boxId)), stateContext, constants.settings.metadata)
-          .map(_ => Unit))
-
-  /**
-    *
-    * @param fb - ergo full block
-    * @return emission box from this block transactions
-    */
-  protected[state] def extractEmissionBox(fb: ErgoFullBlock): Option[ErgoBox] = emissionBoxIdOpt match {
-    case Some(id) =>
-      fb.blockTransactions.txs.view.reverse.find(_.inputs.exists(t => java.util.Arrays.equals(t.boxId, id))) match {
-        case Some(tx) if tx.outputs.head.proposition == constants.genesisEmissionBox.proposition =>
-          tx.outputs.headOption
-        case Some(_) =>
-          log.info(s"Last possible emission box consumed")
-          None
-        case None =>
-          log.warn(s"Emission box not found in block ${fb.encodedId}")
-          boxById(id)
+          .map(_ => Unit)
       }
-    case None =>
-      log.debug("No emission box: emission should be already finished before this block")
-      None
   }
 
-  protected def emissionBoxIdOpt: Option[ADKey] = store.get(ByteArrayWrapper(UtxoState.EmissionBoxIdKey))
-    .map(s => ADKey @@ s.data)
+  protected[state] def extractEmissionBox(fb: ErgoFullBlock): Option[ErgoBox] = {
+    emissionBoxIdOpt match {
+      case Some(id) =>
+        fb.blockTransactions.txs.view.reverse
+          .find(_.inputs.exists(t => java.util.Arrays.equals(t.boxId, id))) match {
+            case Some(tx) if tx.outputs.head.proposition == constants.genesisEmissionBox.proposition =>
+              tx.outputs.headOption
+            case Some(_) =>
+              log.info(s"Last possible emission box consumed")
+              None
+            case _ =>
+              log.info(s"Emission box not found in block ${fb.encodedId}")
+              boxById(id)
+          }
+      case _ => // try to find matching emission box among transactions in block.
+        fb.blockTransactions.txs.view.reverse
+          .find { tx =>
+            tx.outputs.headOption.exists { out =>
+              out.proposition == constants.genesisEmissionBox.proposition &&
+                out.value == constants.emission.remainingCoinsAfterHeight(fb.header.height)
+            }
+          }
+          .map(_.outputs.head)
+          .orElse {
+            log.info("No emission box: emission should be already finished before this block")
+            None
+          }
+    }
+  }
+
+  protected def emissionBoxIdOpt: Option[ADKey] = store.get(ByteArrayWrapper(UtxoState.EmissionBoxIdKey)) match {
```

Comment:
Why is it better then the previous version?
```
 protected def emissionBoxIdOpt: Option[ADKey] = store.get(ByteArrayWrapper(UtxoState.EmissionBoxIdKey))
    .map(s => ADKey @@ s.data)
```
Can you reproduce empty array?

Reply:
@catena2w, yes
https://github.com/ergoplatform/ergo/blob/cd22a12838e940eabeb6a7d89f2df8da4d1e4449/src/main/scala/org/ergoplatform/nodeView/state/UtxoState.scala#L169